Add new option flycheck-shellcheck-infer-shell

jacksonrayhamilton opened the pull request:
Hello,

At my company we want to use ShellCheck to check our scripts, but we also want to support a breadth of OSes.  The choice was made to use Zsh as the default shell because it comes standard in macOS, is much newer than the Bash that comes installed on macOS, and could easily be installed on other OSes.  However since ShellCheck doesn't support Zsh, we are adding the `# shellcheck shell=bash` directive and trying to write mostly-portable code.  Not sure if this sounds unwise but it's where we're at.

What this looks like in practice is that our shell files start like this:

```sh
#!/usr/bin/env zsh
# shellcheck shell=bash
...
```

ShellCheck's documentation indicates that this should be possible: https://github.com/koalaman/shellcheck/wiki/Directive#shell.  And when I test with the `shellcheck` command via CLI, it works.

However, in FlyCheck's checker, currently there are two checks that prevent me from checking such code.

1. The `zsh` shebang will fail the member check for `flycheck-shellcheck-supported-shells`.  I can easily work around that with `(add-to-list 'flycheck-shellcheck-supported-shells 'zsh)`.
2. However, on top of that, the `sh-shellcheck` checker is defined to unconditionally pass the `--shell=zsh` argument in this case, which causes a program error, "Unknown shell: zsh"

The only solution that I'm aware of is for the checker's `:command` to be customizable with a new option that removes the `--shell` arg, allowing for ShellCheck to infer the shell according to the shebang or any `shell` directives in the file.

Let me know if this sounds reasonable to you or if you'd prefer a different approach.  Thanks!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/flycheck/flycheck?pullRequest=2132) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/flycheck/flycheck?pullRequest=2132) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/flycheck/flycheck?pullRequest=2132) it.</sub>

jacksonrayhamilton commented:
I revisited this to try and clean it up before begging for attention.

- Added missing docs for the change (fixing most CI failures)
- Hopefully assigned the new integration test a better identifier than originally
- Updated the original integration test to expect output like the latest stable ShellCheck provides
- Verified that the integration tests pass on my machine

CI still reports some failures in the matrix, but they are the same jobs that failed in the last merged PR: https://github.com/flycheck/flycheck/actions/runs/19103063205/job/54579613912.  So I'm assuming it's not critical that those jobs pass right now.

@bbatsov any feedback?

# Actual ShellCheck output

<img width="1273" height="690" alt="Screenshot 2025-11-15 at 10 46 03 AM" src="https://github.com/user-attachments/assets/87358d0c-cbd5-45b1-98aa-2da3564908cd" />

# Passing ShellCheck integration tests

<img width="1022" height="758" alt="Screenshot 2025-11-15 at 11 16 07 AM" src="https://github.com/user-attachments/assets/0163d88d-2bb1-4bd0-b270-acb51aa7feb3" />

bbatsov commented:
Looks good to me. Thanks!
